### DISCUSSION TITLE
UR inverse kinematics solution rotated 180 degrees

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract#823] [General]
The URInvKin solver is returning results that are 180 degrees rotated about Z compared to the forward kinematics result. I have code that tests KDL, OPW, and UR using the same test. The KDL and OPW work fine, but the UR is incorrect. I've tried fiddling with the parameters, but it seems like there might be an insistency in the solver.

The factory UR parameters have the robot stretched out along -X instead of +X like in ROS.

https://www.universal-robots.com/articles/ur/application-installation/dh-parameters-for-calculations-of-kinematics-and-dynamics/

For their ROS description, they have a 180 degree rotation hacked in to make it match the +X ROS convention:

https://github.com/UniversalRobots/Universal_Robots_ROS2_Description/blob/c4d1724ac32a4f0e3bcfbcfa420f82608b07b17f/urdf/ur_macro.xacro#L270

I am not using a URDF for my code, so I am not entirely sure if this is a problem with my code or with the solver. The failing test is here:

https://github.com/johnwason/rpi_general_robotics_toolbox_py/blob/a5d80df88e89aa95ba4eabfe11a2894972c73900/test/test_tesseract_kinematics.py#L417

I suggest adding a URDF test that calculates forward and inverse kinematics of the UR robot and cross checks between them.

**Comment by johnwason** [comment]
I think the solution to this problem is to add offset transforms before and after the base and tip link for the inverse kinematic solvers that are specified in the yaml files. I am having issues with OPWInvKin as well where I need a slightly different link format than the solver expects.

@Levi-Armstrong do you have a problem with this? What format have you been using for transforms in yaml?

**Reply by Levi-Armstrong** [reply]
Do you have an example of what you are trying to accomplish? The kinematics group handles this for you where you put these types of transforms in the environment. For example the robot urdf goes from base_link to tool0 and then you have a line before base_link called base and a link attached to tool0 called tcp. You can then use the kinematics group solve IK given a transform for tcp in the base frame.

**Reply by johnwason** [reply]
An example is a flange for OPW that does not have Z pointing out of the flange. I'd like to allow my users to have an arbitrary flange link orientation, but right now there isn't any way to specify the offset. Adding another intermediate link is possible, but it may be confusing. It would be easier if there was an offset transform option in the configuration of the plugin. Having offsets for the base and flange would also solve the UR problem since I can specify the transform in the configuration.

**Reply by Levi-Armstrong** [reply]
In the case of the tcp, this can be done by adding a tcp to the srdf. This mimics industrial manipulator controls where the user can specify any number of tcps.

**Reply by johnwason** [reply]
Can the tip_link and base_link specified in the yaml invkin config files be one of the additional tcp?

**Reply by Levi-Armstrong** [reply]
You do specify the base and tip link in the yaml but those must be in the Environment. The tcp is handled by the kinematic group object.

**Reply by Levi-Armstrong** [reply]
The inv kinematics should match the industrial manipulator and the concept of different working frames and tcp are handled by the kinematic group. This mimics the behavior of industrial manipulator controllers so I would be hesitant to add duplicate capability to the ik solver. Now if the ik solver does not match the convention of the actual robot then this should be corrected. In the case of OPW it can be difficult in some cases to make it align which we have discussed about adding the ability to add fixed transforms to OPW Kinematics to solve the issue.

**Reply by johnwason** [reply]
For the UR robot there are two canonical home configurations, stretched out along +X or -X. The controller uses the -X convention, but the ROS configuration is along +X. I'd rather stay with the -X official convention to match what the controller outputs, but the current tesseract solver uses the +X convention. If we have the ability to specify offset transforms for the base and tip links, it would be easy to add in the correction.

**Reply by johnwason** [reply]
OPW solver would benefit from having the ability to specify fixed transforms in the yaml config file.

**Reply by Levi-Armstrong** [reply]
If the controller support multiple conventions, then I would be in favor of updating the yaml to support specifying the different conventions to allow this.

**Reply by gavanderhoorn** [reply]
UR controllers don't support multiple conventions.

It internally aligns its world coordinate frame (which doesn't have that name) with the arm in zero-pose along `x-`.

REP-103 mandates/suggests the *forward* direction for any body-local frame should be `x+`. That's not the case if we use UR's coordinate frame setup.

To rectify this discrepancy, a rotation of `pi` rad around `z+` was introduced. This makes the zero-pose align with `x+`, making everything REP-103 compliant again.

to transform poses such that we can correlate them with the UR-internal world frame is what the `base` frame is for: it's required for `base` to coincide with whatever pose UR's world frame has. If that means it's `pi` rad rotated over `z+` wrt `base_link` (and thus violating REP-103 conventions), so be it.

`base -> tool0` should show the exact same values as the UR TP (after having taken care of the pose display (ie: not use axis-angle)).

To transform a pose "from ROS" to UR would use `base` as the target frame. For the transformation in the opposite direction, go through `base` again.

> I'd rather stay with the -X official convention to match what the controller outputs [..]

Even if Tesseract is ROS-agnostic, I assume many of its users will be either coming from ROS, or looking to use it with URDFs and TF. All of that assumes REP-103 compliance (perhaps even only implicitly). Moving away from REP-103 conventions would seem to make that more complex.

---

Wrt tool frames I'd agree with @Levi-Armstrong: the IK solver should not be responsible for this. Additional frames should be added, and as a convenience, the infrastructure around the IK solver could be setup such that it pre-transforms poses taking that additional frame into account, but the solver proper should not do this.

---

Edit: note there is some discussion about whether the rotation is (still) needed. See the later comments on https://github.com/ros-industrial/universal_robot/pull/505 fi. I've not done any work on those xacro macros the past few years, and they were copied almost verbatim to the ROS 2 description repository/package. You'd have to ask the current maintainers as to the status of it all.

**Reply by Levi-Armstrong** [reply]
I completely forgot about the base versus base_link. This is an error on my end as I should have been setting the yaml base link to base instead of base_link and remove the 180 deg rotation. I will test removing the 180 base rotation and change the yaml base link and make sure all of the unit tests pass.

**Reply by johnwason** [reply]
@Levi-Armstrong it doesn't look like the tip and base links have to be directly in the chain group. I might be able to just add extra links with the correct transforms. Could we maybe add a "quiet" link flag that is ignored by the scene graph when looking for tip and base links? It could also be used to mark frames that are for fiducial marks or sensors so they are ignored by the solvers.

**Reply by Levi-Armstrong** [reply]
@gavanderhoorn So I start to look into removing the 180deg base rotation but it appears that the urdf does not have the base link. Am I missing something?

**Reply by gavanderhoorn** [reply]
I'm not sure I follow.

Which URDF are you referring to?

**Reply by Levi-Armstrong** [reply]
Here https://github.com/ros-industrial/universal_robot/tree/kinetic-devel/ur_e_description/urdf

**Reply by gavanderhoorn** [reply]
That's still just a directory, not a specific URDF or `.xacro`.

But https://github.com/ros-industrial/universal_robot/blob/06d8b9e2f5f86aa54f9f2845f11edbc84e2f951e/ur_e_description/urdf/ur10e.urdf.xacro#L74 seems to be the `base_link`.

but `melodic-devel-staging` would be more recent, and that has it here: https://github.com/ros-industrial/universal_robot/blob/addbe3fc2b8f4fd3afc65d2c8f9a5380324e04ba/ur_description/urdf/inc/ur_macro.xacro#L79.

**Reply by Levi-Armstrong** [reply]
My understanding is that it should have a link called "base" where if you queried from base to tcp it should match the robot controller.

**Reply by Levi-Armstrong** [reply]
In the case of abb it has the following
``` xml
    <!-- ROS base_link to ABB World Coordinates transform -->
    <link name="${prefix}base" />
    <joint name="${prefix}base_link-base" type="fixed">
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <parent link="${prefix}base_link"/>
      <child link="${prefix}base"/>
    </joint>
```

**Reply by gavanderhoorn** [reply]
Ah. You didn't format it as `base` link, so I got confused.

That's here: https://github.com/ros-industrial/universal_robot/blob/addbe3fc2b8f4fd3afc65d2c8f9a5380324e04ba/ur_description/urdf/inc/ur_macro.xacro#L306-L317.

**Reply by Levi-Armstrong** [reply]
Okay, so I have been using the incorrect xacro file when testing. I will see if I can fix the ik solver so it does not have the 180 degree base rotation.

**Reply by gavanderhoorn** [reply]
Well the chain would be identical to the one in `kinetic-devel`. It also has the `base` link and the rotation.

**Comment by Levi-Armstrong** [comment]
> I am not using a URDF for my code, so I am not entirely sure if this is a problem with my code or with the solver. The failing test is here:

There are test in tesseract for the UR which compare the IK to the Forward so I assume the issue is not within tesseract.  Now I also had to add a base offset in the inverse kinematics solver for the UR to have it align with the URDF's within ur_descriptions. If you look in ur_inv_kin.cpp line 262 you will 180deg rotation around Z axis.
